### ISSUE TITLE
Why must SameSite=none?

### ISSUE THREAD

**Issue by anderspitman** [fedidcg/FedCM#587]
I've searched the repo and read through https://github.com/fedidcg/FedCM/issues/212 and https://github.com/fedidcg/FedCM/issues/248 but I'm still confused about the SameSite=none requirement.

[LastLogin](https://lastlogin.io) currently uses SameSite=lax for all cookies. Changing to none potentially opens up several vulnerabilities that need to be carefully considered. I get some protection from CORS, but there are so many edge cases that things can slip through with "simple requests" etc. I could also add a tiny sentinel cookie set with SameSite=lax and then require both cookies for all endpoints except FedCM, but again that requires me to get it right everywhere, which is dangerous.

Overall, why is this necessary in the first place. As @samuelgoto mentions [here](https://github.com/fedidcg/FedCM/issues/248#issuecomment-1113942590) it seems to me that FedCM already is secure enough to be trusted sending lax cookies.

**Comment by sebadob** [comment]
That's a big pain point for me as well and I would actually never deploy something to prod with this setting.  

I think the browser should just sent even host only cookies to the IdP in that case and since it's behind the scenes anyway, I don't see an issue with doing that. The other way around, setting `SameSite=None` is something I would never do these days, as it would be like 3 steps back in terms of security.

**Comment by anderspitman** [comment]
@samuelgoto any additional info on this?

**Comment by samuelgoto** [comment]
> @samuelgoto any additional info on this?

We are working with our privacy and security reviewers to figure out what we can do about this. Will report back.

**Comment by samuelgoto** [comment]
> We are working with our privacy and security reviewers to figure out what we can do about this. Will report back.

@johannhof can you help us understand why `SameSite=None` is required and `SameSite=Lax` wouldn't suffice?

**Comment by cbiesinger** [comment]
@bvandersloot-mozilla @annevk fyi

**Comment by johannhof** [comment]
I think I understand your concern and that you consider requests initiated by FedCM to meet a higher bar than "regular" cross-site fetch requests, however, let's look at the available options and [their current definitions](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis/#section-4.1.2.7):

- SameSite=Strict, for same-site requests (which e.g. accounts endpoint is clearly not, as it's triggered in a cross-site context)
- SameSite=Lax, for same-site requests and cross-site top-level navigations (which this is also not)
- SameSite=None

Of course, we can revisit and change things if they don't make sense. This was discussed in https://github.com/fedidcg/FedCM/issues/248 and the suggestion back then was to add a `SameSite=Federated` attribute, which probably doesn't work due to backwards/cross-browser compat issues.

We could consider broadening the definition of `SameSite=Lax` to cover things like browser-UI triggered requests, but we'd need to make sure that we're not inadvertently opening up an attack surface for anyone. If you were previously using the above definition to protect your endpoints you may be surprised to learn that there's a new cross-site way to trigger `SameSite=Lax` cookies.

cc @arturjanc for thoughts

**Comment by johannhof** [comment]
@dcthetall FYI ^

**Comment by anderspitman** [comment]
Thanks @johannhof.

I think the problem is that FedCM requests feel like a new thing. Is there any precedent for requests that are sent by the browser but which the origin has no direct control over the method, content, or response (only the URI can be controlled by the RP)?

I can see why you technically consider them to be cross-site requests since they wouldn't be fired without visiting an RP, but do they really fit the description of what we've always considered cross-site on the web up to this point? The term is generally used in conversations involving CSRF security concerns and privacy, and I haven't been able to come up with any scenarios where FedCM requests would introduce a security vulnerability (caveat: I am not a security expert). The endpoints are highly prescribed and the results aren't directly accessible to the RP.

On the flip side, forcing all cookies to be SameSite=none for FedCM introduces real, tangible security issues.

**Comment by johannhof** [comment]
Yeah, I understand the negative security implications this would cause for IdPs in the mid term (that is, until 3rd party cookies are blocked by default without workarounds in all browsers) and I agree it's not a great outcome.

I'm honestly torn on this issue. I'd love to hear from some of the tagged folks (@arturjanc @annevk @bvandersloot-mozilla) regarding their thoughts on a possible inclusion of FedCM requests in `SameSite=Lax` rules.

**Comment by sebadob** [comment]
I think it would make a lot of sense to even be able to set host only cookies, even with `__Host-*` and have the FedCM send them anyway, even it is is technically cross origin. It would not be security issue or against any "rules", because the FedCM API operates behind the scenes from inside the browser and not on the user-facing side, and can be trusted.  

On the other hand, if we would set host only or samesite cookies, they are restricted to the IdPs domain / host and would not leak anywhere else, even when the FedCM is not used.

**Comment by sebadob** [comment]
@johannhof don't get me wrong, I would never want this to be a global change, only for these specific endpoints where we actually need the cookies during the FedCM flow. Basically like you suggested. 

The `SameSite` attribute is one of the best defenses we have against CSRF attacks. I would never want to give that up just to support some new auth flow, even if it provides a really nice UX, and rather stick to OAuth / OIDC on its own.

**Comment by annevk** [comment]
I don't understand the discussion about `__Host-` cookies. That's a syntactic mechanism to enforce a couple of attributes of a cookie. It's unrelated to `SameSite` and can always be used. Are you suggesting to change the semantics in some way? If so, please explain by way of an example involving the multiple parties. But also, if it's not related to `SameSite` it should probably be its own issue.

Creating a new avenue for `SameSite=Lax` exploits makes me rather nervous. I guess to some extent the question here is whether the "speedbumps" for attackers as 6265bis puts it are equivalent.

**Comment by sebadob** [comment]
> I don't understand the discussion about __Host- cookies. That's a syntactic mechanism to enforce a couple of attributes of a cookie. It's unrelated to SameSite and can always be used. Are you suggesting to change the semantics in some way? If so, please explain by way of an example involving the multiple parties. But also, if it's not related to SameSite it should probably be its own issue.

Sorry the `__Host-` does not belong here, that's true, my bad.

**Comment by anderspitman** [comment]
I appreciate the time being taken to seriously consider this concern. Honestly the situation makes me a bit uneasy. Just because it seems safer to relax SameSite=lax than to require SameSite=none doesn't necessarily mean that's the case. My instincts tell me this should be treated as a new thing. The SameSite=federated suggestion should work. But if people who know more about security than me (not a high bar) sign off on SameSite=lax being relaxed, I won't complain.

**Comment by bc-pi** [comment]
I tend to agree that a `SameSite=None` requirement for cookies to be sent with requests to both accounts and identity assertion endpoints will make IDP side implementation of FedCM more cumbersome and/or prone to introducing other vulnerabilities. 

My intuition is that the FedCM requests are something fundamentally different and it'd be appropriate/okay/safe to send `SameSite=Lax` cookies or even ignore the SameSite directive. These FedCM requests are sent to new endpoints as part of interacting with this new API that's intended to facilitate federated identity style functionality in a post 3rd party cookie world. And they'll be sent with a [Sec-Fetch-Dest](https://fedidcg.github.io/FedCM/#sec-fetch-dest-header) header field. They are definitely not top-level navigation but they aren't typical subresource requests either. 

I'd note that the [document](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis) to which @johannhof previously referred is is [currently in Working Group Last Call](https://lists.w3.org/Archives/Public/ietf-http-wg/2024AprJun/0024.html) in the IETF's HTTP WG. Would it be worthwhile to provide feedback about this here issue and try and get a carve-out or allowance of some sort for these unique kinds of requests? Is anyone aware of similar platform originating requests and if or how they are handling similar considerations?

**Comment by cbiesinger** [comment]
Implementation-wise, Chrome used to send SameSite=Strict cookies as well, however there was strong opposition to that so we ended up with what we do now (SameSite=None only).

**Comment by samuelgoto** [comment]
> Implementation-wise, Chrome used to send SameSite=Strict cookies as well, however there was strong opposition to that so we ended up with what we do now (SameSite=None only).

@cbiesinger would it be equally trivial / easy for Chrome to send `SameSite=Lax` as well?

**Comment by cbiesinger** [comment]
> > Implementation-wise, Chrome used to send SameSite=Strict cookies as well, however there was strong opposition to that so we ended up with what we do now (SameSite=None only).
> 
> @cbiesinger would it be equally trivial / easy for Chrome to send `SameSite=Lax` as well?

I am currently inquiring about that

**Comment by cbiesinger** [comment]
> But this means at the same time, that any javascript code in the internet can make a request to your IdP and will simply receive a token (as long as the user is currently logged in), if it just sets the correct headers, which are no secret.

The FedCM endpoints in particular, which I think you are talking about (?), can/should verify that the `sec-fetch-dest` header is set to `webidentity` to avoid this issue. sec- headers can not be set by JS code.

**Comment by bc-pi** [comment]
@cbiesinger mentioned this over in #212 but I think it bears repeating here - the the last paragraph in https://fedidcg.github.io/FedCM/#browser-api (copied below because content at github.io isn't necessarily stable) actually contradicts the behavior to only send SameSite=None designated cookies. 
 
> For fetches that are sent with cookies, unpartitioned cookies are included, as if the resource was loaded as a same-origin request, e.g. regardless of the [SameSite](https://httpwg.org/http-extensions/draft-ietf-httpbis-rfc6265bis.html#name-the-samesite-attribute-2) value (which is used when a resource loaded as a third-party, not first-party). This makes it easy for an [IDP](https://fedidcg.github.io/FedCM/#idp) to adopt the FedCM API. It doesn’t introduce security issues on the API because the [RP](https://fedidcg.github.io/FedCM/#rp) cannot inspect the results from the fetches in any way.

**Comment by bc-pi** [comment]
> Implementation-wise, Chrome used to send SameSite=Strict cookies as well, however there was strong opposition to that so we ended up with what we do now (SameSite=None only).

Does that imply that the google account cookies are `SameSite=None`?

**Comment by johannhof** [comment]
> I'd note that the [document](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis) to which @johannhof previously referred is is [currently in Working Group Last Call](https://lists.w3.org/Archives/Public/ietf-http-wg/2024AprJun/0024.html) in the IETF's HTTP WG. Would it be worthwhile to provide feedback about this here issue and try and get a carve-out or allowance of some sort for these unique kinds of requests? Is anyone aware of similar platform originating requests and if or how they are handling similar considerations?

If we agree on the `SameSite=Lax` idea we should specify it and eventually standardize but it seems a bit premature to encode this in an actual standards document without multi-implementer support. After 6265bis It'll also get easier to just specify this in FedCM and/or Fetch going forward, without going through the RFC, as part of the planned [cookie layering](https://github.com/httpwg/http-extensions/issues/2084) changes.

Note that browsers can make this change without requiring standardization, the standard usually follows implementation.

**Comment by annevk** [comment]
Before we can reach agreement I think someone has do the work on explaining the limited protections that exist around SameSite=Lax today and how this feature doesn't make the situation worse for existing SameSite=Lax cookies.

Also, standardization as a concept very much is inclusive of discussions such as this one so I can't really agree with your statement there.

**Comment by bc-pi** [comment]
> Before we can reach agreement I think someone has do the work on explaining the limited protections that exist around SameSite=Lax today and how this feature doesn't make the situation worse for existing SameSite=Lax cookies.

As best I understand and can summarize things here, this would mean attaching SameSite=Lax cookies only to requests initiated via this new browser API to a few new endpoints that are required to enforce the presence of a `Sec-Fetch-Dest: webidentity` header and don't directly expose the response content. I'll not claim perfection but there's CSRF and some XSS protections there. I don't see how it exacerbates the situation for existing SameSite=Lax cookies. But it would prevent IDP implementations wanting to adopt FedCM from "downgrading" their existing SameSite=Lax cookies to SameSite=None and/or doing cumbersome and potentially error-prone workarounds.  

> Also, standardization as a concept very much is inclusive of discussions such as this one so I can't really agree with your statement there.

Yeah, standardization needs to be inclusive and more than just documenting implementation.

**Comment by johannhof** [comment]
Yeah, to be clear I wasn't trying to say "we can't do this because of some clause in 6265bis". We need to threat model how this impacts developers relying on current SameSite definitions, which happen to be specified in 6265bis. If we do this analysis and get implementer buy-in then we'll make whatever changes are needed in FedCM and the Cookies RFC.

On Chrome side I haven't seen anyone strongly object to this from a security perspective and there seems to be significant upside for developers, so I think we're tentatively positive on sending `SameSite=Lax`. @mikewest any concerns?

**Comment by cbiesinger** [comment]
Chrome Canary now supports sending SameSite=Lax cookies for credentialed FedCM requests behind a flag, `--enable-features=FedCmSameSiteLax` (needs to be specified on the commandline, there is no chrome://flags entry)

@bvandersloot-mozilla let me attempt to summarize my understanding here:
SameSite=None cookies get sent on all cross-site requests, including fetch(), `<img>`, etc
Lax cookies only get sent for navigations (e.g. `<iframe>`)
So today, you can get third-party signin functionality by using an iframe from the IDP domain while using Lax cookies.

But if FedCM requires None, that means you have to allow your cookies to be sent in a lot more cases, which is undesirable.

Can someone confirm whether I summarized the issue correctly?

**Comment by bvandersloot-mozilla** [comment]
> Chrome Canary now (starting in 128.0.6589.0) supports sending SameSite=Lax cookies for credentialed FedCM requests behind a flag, `--enable-features=FedCmSameSiteLax` (needs to be specified on the commandline, there is no chrome://flags entry)
> 
> @bvandersloot-mozilla let me attempt to summarize my understanding here: Lax cookies only get sent for navigations (e.g. `<iframe>`)

Not quite, Lax cookies are more strict than that. They are only sent on top-level navigations, using safe methods. There are additional relaxations in place for cookies that are Lax because they have no attribute set, but those are sort-of-bad to rely upon and are an artifact of failing to migrate the default SameSite attribute fully.

> So today, you can get third-party signin functionality by using an iframe from the IDP domain while using Lax cookies.
I'm not sure how these things are working now- maybe it is the SameSite-Lax-by-default exceptions.

> But if FedCM requires None, that means you have to allow your cookies to be sent in a lot more cases, which is undesirable.

Agreed that this is probably the case, but if that is because the previous behavior was using the SameSite-Lax-by-default exceptions we should find a better answer for them.

**Comment by cbiesinger** [comment]
Hi @anderspitman, could you explain how you use SameSite=Lax cookies today, especially with respect to Ben's last comment? Do you only use redirects today?

**Comment by anderspitman** [comment]
@cbiesinger sorry for the late response. Yes, I'm only using redirects currently.

**Comment by anderspitman** [comment]
@samuelgoto didn't realize I might not be blocked on this. I'll give it a shot.

**Comment by anderspitman** [comment]
@samuelgoto @cbiesinger unfortunately this doesn't seem to be working. Running on version `130.0.6683.2` on Linux. I have an idea for a workaround involving extra cookies but not 100% sure that will work yet, and I'm worried about introducing subtle vulnerabilities.

**Comment by anderspitman** [comment]
I've implemented my workaround for prod LastLogin. The way it works is I set a simple lax cookie for all requests, and made my login cookies SameSite=none. I then gated them behind a check to verify the lax cookie is present for all requests except FedCM /accounts and /id-assertion.

I would appreciate feedback from those in this thread to make sure I'm not missing any subtle security holes.

**Comment by npm1** [comment]
I am curious regarding why it did not work for you. Did you enable the feature, restarted Chrome, and tried it out? Did you not get the Lax cookies or what didn't work?

**Comment by anderspitman** [comment]
@npm1 with the feature enabled looks like Chrome is sending the cookies for /accounts, but not /id-assertion. Without the flag it doesn't send them for either.

**Comment by npm1** [comment]
Ok thanks! I pointed this out in https://issues.chromium.org/issues/350944661 although Christian is out till Wednesday

**Comment by anderspitman** [comment]
Thanks for passing that on. And no worries, I'm *pretty* sure my workaround is secure and there's no rush to get this implemented on my end.

It definitely would still be very useful in the future for added security and for ease for new implementers. I know @sebadob has stated he wouldn't deploy FedCM for Rauthy unless lax cookies can be used. Not sure if his position has changed.

**Comment by samuelgoto** [comment]
>  I'm pretty sure my workaround is secure 

Can you expand on what workaround you used?

**Comment by anderspitman** [comment]
I mentioned it here: https://github.com/w3c-fedid/FedCM/issues/587#issuecomment-2336489848

I can give more details if that's confusing. Basically since we have to use SameSite=None, the browser is always sending the login cookies, even cross-domain. So I added another SameSite=Lax cookie that acts as a "key" that turns on the login cookie. All endpoints require that key to be present except for fedcm/accounts and fedcm/id-assertion

**Comment by cbiesinger** [comment]
So it turns out that making SameSite=Lax is harder than it initially seemed. Since the ID assertion request is a POST request, and POST requests do not receive SameSite=Lax cookies, it is not enough to treat these requests as top-level requests.

We also can't treat ID assertion requests as same-site requests due to the CORS requirement.

So this might require discussions involving the fetch spec and/or cookie layering...

It seems @anderspitman has a workaround for now, so am not planning to work on this issue for now.

**Comment by thhck** [comment]
> I can give more details if that's confusing. Basically since we have to use SameSite=None, the browser is always sending the login cookies, even cross-domain. So I added another SameSite=Lax cookie that acts as a "key" that turns on the login cookie. All endpoints require that key to be present except for /fedcm/accounts and /fedcm/id-assertion


I’d appreciate any additional details about how you implemented this if it's not too much trouble. You duplicated the session cookie ( one with `Lax` the other with `None` ) and check if both exist for all authenticated endpoint expect for the fedcm ones ?

**Comment by anderspitman** [comment]
> > I can give more details if that's confusing. Basically since we have to use SameSite=None, the browser is always sending the login cookies, even cross-domain. So I added another SameSite=Lax cookie that acts as a "key" that turns on the login cookie. All endpoints require that key to be present except for /fedcm/accounts and /fedcm/id-assertion
> 
> I’d appreciate any additional details about how you implemented this if it's not too much trouble. You duplicated the session cookie ( one with `Lax` the other with `None` ) and check if both exist for all authenticated endpoint expect for the fedcm ones ?

Almost. The Lax cookie doesn't actually carry any authentication data. It's just a sentinel that's set for any browser that visits the site. If that cookie is present I know it's not a cross-site request so it's safe to check the actual auth cookie. For some of the FedCM endpoints I trust the auth cookie without requiring the Lax sentinel to be resent.

**Comment by npm1** [comment]
Is this still a blocker for anyone?

**Comment by thhck** [comment]
In my case yes.
In absolute terms I guess I could implement the workaround, but in practical terms it's too much of a hassle

**Comment by bc-pi** [comment]
> Is this still a blocker for anyone?

nothing has changed from the prior comments I've made, such as https://github.com/w3c-fedid/FedCM/issues/587#issuecomment-2130298419

**Comment by npm1** [comment]
Well, did you try https://github.com/w3c-fedid/FedCM/issues/587#issuecomment-2223335737? Only @anderspitman responded and they are unblocked as far as I understand. @cbiesinger mentioned it currently only works for the accounts endpoint but not for the ID assertion endpoint. I'm also not sure about what's the most correct behavior here from a security perspective

**Comment by anderspitman** [comment]
I'm unblocked but unsure my solution is secure. It's simple so hopefully I'm not missing anything but no one has commented on my approach one way or the other.

**Comment by hlflanagan** [comment]
Discussed during the FedID WG/CG meeting 22 April 2025 <https://github.com/w3c-fedid/meetings/blob/main/2025/2025-04-22-FedCM-agenda.md>

**Comment by cbiesinger** [comment]
> Before we can reach agreement I think someone has do the work on explaining the limited protections that exist around SameSite=Lax today and how this feature doesn't make the situation worse for existing SameSite=Lax cookies.
> 
> Also, standardization as a concept very much is inclusive of discussions such as this one so I can't really agree with your statement there.

Here is my attempt at explaining how this proposal does not make things worse for SameSite=Lax:

# Background

Various IDPs today use redirect flows to let users sign into an RP. These IDPs want to only send cookies to these top-level page loads and redirects, so they use SameSite=Lax.

FedCM currently only sends SameSite=None cookies (see extended previous discussion in #320 and at f2f meetings). IDPs want to adopt FedCM, but not loosen their security posture by sending their login cookies as SameSite=None.

# Proposal

To solve this, the proposal is to also send Lax cookies in credentialed fedcm requests (i.e. accounts endpoint, ID assertion endpoint, and disconnect endpoint)

# Security

IDPs need to specifically opt in to FedCM by having a (/.well-known/web-identity)[https://w3c-fedid.github.io/FedCM/#idp-api-well-known] file on their eTLD+1 host which specifies either the accounts endpoint or the FedCM config file (which lists the accounts endpoint, id assertion endpoint, and disconnect endpoint, among others not relevant here)

Because IDPs specifically need to opt in to these endpoints the risk of undesired cookie leaks is very limited. In fact, there can be only one endpoint across an eTLD+1 that will receive such cookies before specific user interaction (the accounts endpoint), at least for a new user to a specific (RP, IDP) pair (we allow automatic reauthentication for returning users, which would then send these cookies also to the ID assertion endpoint. Also, for a returning user, an RP can call disconnect, which would make a credentialed request to the disconnect endpoint).

To be explicit: If an IDP does not have a /.well-known/webidentity file on their eTLD+1, this proposal will not change when SameSite=Lax cookies are sent.

**Comment by johannhof** [comment]
(Which was added as part of cookie layering)

**Comment by cbiesinger** [comment]
> Starting in tomorrow's Chrome Canary, you can test this using --enable-features=SendSameSiteLaxForFedCM. Sorry, this is not listed in chrome://flags, but will work from the commandline

@anderspitman / others, I am still interested in hearing if this is working for you!

**Comment by anderspitman** [comment]
Sorry @cbiesinger busy times for me, but I'll see if I can carve out a couple hours to play with this in the next 2 weeks. Are there any looming deadlines on your end where it would be nice to get this tested?

**Comment by will-bartlett** [comment]
I stumbled across this today. +1 for this being a problem. Big picture - SameSite is used to protect websites against unexpected requests. FedCM requests are never unexpected, because the browser will only make them cross origin if the origin has configured a well known file.
